File: packages/ts/tests/integration/binding/path-table-no-ignore.test.ts
import { mkdir, mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { DirSQL } from "dirsql";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

// Path-table scans respect .gitignore by default (#742); `noIgnore: true`
// opts back in to the ignored files while the built-in floor (node_modules,
// .git) still applies. The behavior lives in the Rust core -- these tests
// prove the constructor option crosses the napi boundary.
describe("path-table gitignore opt-out (#746)", () => {
  let dir: string;

  beforeEach(async () => {
    dir = await mkdtemp(join(tmpdir(), "dirsql-746-"));
    await mkdir(join(dir, "dist"), { recursive: true });
    await mkdir(join(dir, "src"), { recursive: true });
    await mkdir(join(dir, "node_modules", "pkg"), { recursive: true });
    await writeFile(join(dir, ".gitignore"), "dist/\n*.log\n");
    await writeFile(join(dir, "dist", "bundle.js"), "js");
    await writeFile(join(dir, "src", "app.js"), "js");
    await writeFile(join(dir, "debug.log"), "log");
    await writeFile(join(dir, "node_modules", "pkg", "index.js"), "js");
  });

  afterEach(async () => {
    await rm(dir, { recursive: true, force: true });
  });

  const paths = async (db: DirSQL) => {
    const rows = await db.query("SELECT path FROM './'");
    return rows.map((r) => r.path).sort();
  };

  it("excludes gitignored files by default", async () => {
    const scanned = await paths(new DirSQL({ root: dir }));

    expect(scanned).not.toContain("dist/bundle.js");
    expect(scanned).not.toContain("debug.log");
    expect(scanned).toContain("src/app.js");
  });

  it("restores gitignored files under noIgnore but keeps the built-in floor", async () => {
    const scanned = await paths(new DirSQL({ root: dir, noIgnore: true }));

    expect(scanned).toContain("dist/bundle.js");
    expect(scanned).toContain("debug.log");
    expect(scanned).not.toContain("node_modules/pkg/index.js");
  });
});

File: packages/python/tests/integration/binding/path_table_no_ignore_test.py
"""Binding-tier tests (real core, real fs) for the ``no_ignore`` opt-out.

Path-table scans respect ``.gitignore`` by default; ``no_ignore=True``
restores the full walk. The behavior lives in the shared Rust core -- these
tests prove the constructor opt-out crosses the PyO3 boundary.
"""

import pytest

from dirsql import DirSQL


@pytest.fixture
def gitignored_dir(tmp_path):
    (tmp_path / ".gitignore").write_text("ignored.md\n", encoding="utf-8")
    (tmp_path / "kept.md").write_text("kept", encoding="utf-8")
    (tmp_path / "ignored.md").write_text("ignored", encoding="utf-8")
    return str(tmp_path)


def describe_no_ignore():
    @pytest.mark.asyncio
    async def it_excludes_gitignored_files_by_default(gitignored_dir):
        rows = await DirSQL(gitignored_dir).query("SELECT path FROM './'")

        assert sorted(r["path"] for r in rows) == [".gitignore", "kept.md"]

    @pytest.mark.asyncio
    async def it_returns_gitignored_files_with_no_ignore(gitignored_dir):
        db = DirSQL(gitignored_dir, no_ignore=True)

        rows = await db.query("SELECT path FROM './'")

        assert sorted(r["path"] for r in rows) == [
            ".gitignore",
            "ignored.md",
            "kept.md",
        ]

    @pytest.mark.asyncio
    async def it_keeps_the_built_in_floor_under_no_ignore(gitignored_dir, tmp_path):
        (tmp_path / "node_modules").mkdir()
        (tmp_path / "node_modules" / "dep.js").write_text("x", encoding="utf-8")
        db = DirSQL(gitignored_dir, no_ignore=True)

        rows = await db.query("SELECT path FROM './'")

        assert "node_modules/dep.js" not in [r["path"] for r in rows]
